airbyte-integrations/bases/debezium-v1-9-6/src/main/java/io/airbyte/integrations/debezium/internals/DebeziumRecordIterator.java

```diff
@@ -71,34 +83,68 @@ protected ChangeEvent<String, String> computeNext() {
     while (!MoreBooleans.isTruthy(publisherStatusSupplier.get()) || !queue.isEmpty()) {
       final ChangeEvent<String, String> next;
       try {
-        final Duration waitTime = receivedFirstRecord ? SUBSEQUENT_RECORD_WAIT_TIME : firstRecordWaitTime;
+        // #18987: waitTime here is largely meaningless with heartbeats but leaving for backward
+        // compatibility
+        // for connectors not implementing heartbeat yet (MySql, MSSql)
+        final Duration waitTime = receivedFirstRecord ? SUBSEQUENT_RECORD_WAIT_TIME : this.firstRecordWaitTime;
         next = queue.poll(waitTime.getSeconds(), TimeUnit.SECONDS);
       } catch (final InterruptedException e) {
         throw new RuntimeException(e);
       }
 
       // if within the timeout, the consumer could not get a record, it is time to tell the producer to
       // shutdown.
+      // #18987: Noticed in testing that it's possible for DBZ to be stuck "Searching for WAL resume
+      // position"
+      // when no changes exist. In that case queue will pop after timeout with null value for next
       if (next == null) {
-        LOGGER.info("Closing cause next is returned as null");
+        LOGGER.info("Closing: queue returned null event");
         requestClose();
         LOGGER.info("no record found. polling again.");
         continue;
       }
 
+      // check if heartbeat and read hearbeat position
+      LOGGER.debug("checking heartbeat lsn for: {}", next);
+      final Long heartbeatPos = targetPosition.getHeartbeatPosition(next);
+      if (heartbeatPos != null) {
+        // wrap up sync if heartbeat position crossed the target OR heartbeat position hasn't changed for
+        // too long
+        if (targetPosition.reachedTargetPosition(heartbeatPos)
+            || (heartbeatPos.equals(this.lastHeartbeatPosition) && heartbeatPosNotChanging())) {
+          LOGGER.info("Closing: Heartbeat indicates sync is done");
+          requestClose();
+        }
+        if (!heartbeatPos.equals(this.lastHeartbeatPosition)) {
+          this.tsLastHeartbeat = LocalDateTime.now();
+          this.lastHeartbeatPosition = heartbeatPos;
+        }
+        continue;
+      }
+
       final JsonNode eventAsJson = Jsons.deserialize(next.value());
       hasSnapshotFinished = hasSnapshotFinished(eventAsJson);
 
       // if the last record matches the target file position, it is time to tell the producer to shutdown.
+
       if (!signalledClose && shouldSignalClose(eventAsJson)) {
+        LOGGER.info("Closing: Change event reached target position");
         requestClose();
       }
+      this.tsLastHeartbeat = null;
+      this.lastHeartbeatPosition = null;
       receivedFirstRecord = true;
       return next;
     }
     return endOfData();
   }
 
+  private boolean heartbeatPosNotChanging() {
+    final Duration tbt = Duration.between(this.tsLastHeartbeat, LocalDateTime.now());
+    LOGGER.debug("Time since last hb_pos change {}s", tbt.toSeconds());
```

Comment:
Could you point me towards the setting where we define how frequently we receive heartbeats?